src/java.base/macosx/native/libjava/java_props_macosx.c

```diff
@@ -229,33 +230,43 @@ void setOSNameAndVersion(java_props_t *sprops) {
     NSString *nsVerStr = NULL;
     char* osVersionCStr = NULL;
     NSOperatingSystemVersion osVer = [[NSProcessInfo processInfo] operatingSystemVersion];
-    // Copy out the char* if running on version other than 10.16 Mac OS (10.16 == 11.x)
-    // or explicitly requesting version compatibility
-    if (!((long)osVer.majorVersion == 10 && (long)osVer.minorVersion >= 16) ||
-            (getenv("SYSTEM_VERSION_COMPAT") != NULL)) {
-        if (osVer.patchVersion == 0) { // Omit trailing ".0"
+    // Some macOS versions require special handling. for example, when the NSOperatingSystemVersion
+    // reports 10.16 as the version then it should be treated as 11. Similarly, when it reports 16.0
+    // as the version then it should be treated as 26.
+    // If the SYSTEM_VERSION_COMPAT environment variable (a macOS construct) is set to 1, then we
+    // don't do any special handling for any versions and just literally use the value that
+    // NSOperatingSystemVersion reports.
+    const char* envVal = getenv("SYSTEM_VERSION_COMPAT");
+    const bool versionCompatEnabled = envVal != NULL && strncmp(envVal, "1", 1) == 0;
+    const bool requiresSpecialHandling = ((long) osVer.majorVersion == 10 && (long) osVer.minorVersion >= 16)
+                                         || ((long) osVer.majorVersion == 16 && (long) osVer.minorVersion >= 0);
```

Comment:
Since we know Apple jumped from 15 --> 26, would it be more future-proof to change this test to something like this?

```
    osVer.majorVersion >= 16 && osVer.majorVersion < 26 
```

This would allow us to work when macOS 27 comes out, in case it reports itself as "17.0" if we don't update the version of Xcode before then.

I can also see the argument for leaving it as you have done. And my guess is that Apple will report macOS 27 as 27 regardless, so this probably doesn't matter.

Reply:
Hello Kevin,

> I can also see the argument for leaving it as you have done.

I gave your suggestion some thought. I think it would be simpler (and less confusing) to leave that check in the current form and update it as and when needed depending on whether macOS does report future versions as two different values.